quickwit-storage/src/split.rs

```diff
@@ -0,0 +1,379 @@
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io::{self, Cursor, ErrorKind, SeekFrom, Write};
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+
+use async_trait::async_trait;
+use futures::StreamExt;
+use rusoto_core::ByteStream;
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+use tokio_util::io::ReaderStream;
+
+use crate::{BundleStorageFileOffsets, PutPayload};
+
+trait DataSource: tokio::io::AsyncRead + Send + Sync + Unpin {}
+impl<T: tokio::io::AsyncRead + Send + Sync + Unpin> DataSource for T {}
+
+/// SplitStreamer to Stream the split via the PutPayload.
+/// Bundles together multiple files into a single file
+/// with some metadata
+#[derive(Clone)]
+pub struct SplitStreamer {
+    /// Total split size
+    pub split_size: usize,
+    /// Files and their target range in the split bundle.
+    pub files_and_offsets: Vec<(PathBuf, Range<usize>)>,
+    /// bytes range of the footer (hotcache + bundle metadata)
+    pub footer_range: Range<usize>,
+    /// data of the footer, containng
+    /// [bundle metadata, bundle metadata len 8byte, hotcache, hotcache len 8 bytes]
+    pub footer_bytes: Vec<u8>,
+    /// The end offset of the data provider [File, .. File, SplitFooter]
+    pub block_offsets: Vec<Range<usize>>,
+}
+
+impl SplitStreamer {
+    fn get_start_and_end_index(&self, range: Range<usize>) -> io::Result<(usize, usize)> {
+        get_start_and_end_index(&self.block_offsets, range)
+    }
+}
+
+#[async_trait]
+impl PutPayload for SplitStreamer {
+    async fn len(&self) -> io::Result<u64> {
+        Ok(self.split_size as u64)
+    }
+
+    async fn range_byte_stream(&self, range: Range<u64>) -> io::Result<ByteStream> {
+        let (start_index, end_index) =
+            self.get_start_and_end_index(range.start as usize..range.end as usize)?;
+
+        let start_file_offset = range.start - self.block_offsets[start_index].start as u64;
+
+        // If we have skip and take in the same block, we need to take the skip into account for
+        // the take calculation
+        let same_block_skip = if start_index == end_index {
+            start_file_offset
+        } else {
+            0
+        };
+        let last_block_take =
+            range.end - self.block_offsets[end_index].start as u64 - same_block_skip;
+
+        // build data_sources, apply seek in first
+        let mut datasources: Vec<Box<dyn DataSource>> = vec![];
+        let files_len = self.files_and_offsets.len();
+        let mut first_ds = true;
+
+        if start_index < files_len {
```

Comment:
It seems like a lot of the complexity is coming from the fact that you need to special case the footer bytes.

I think a no brainer to simplify this code is to introduce a private method `get_data_source` to this struct that takes a `chunk id, and a range`.
The function should start with a defensive assert checking if the range matches the given chunk, and return the right data source.

The current method should then be trivial.

Reply:
I added the method, because of the seek there is still some complexity, but it's improved